### PR TITLE
Added performance_regression_row_level_repair_test.py and its utils

### PR DESCRIPTION
**Pull request by yarongilor**
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by yarongilor** [comment]
An example test run OK with result output of:
```
16:51:41 < t:2019-08-28 13:51:41,839 f:results_analyze.py l:304  c:sdcm.results_analyze p:INFO  > Performance result for: repair_runtime_large_partitions is: 195.017749071. (average statistics deviation limit is: 133.342990065
16:51:41 < t:2019-08-28 13:51:41,839 f:results_analyze.py l:310  c:sdcm.results_analyze p:INFO  > Performance average of 3 results for: repair_runtime_large_partitions on version: 3.1.0.rc2 is: 28.1500016848
16:51:41 < t:2019-08-28 13:51:41,839 f:results_analyze.py l:310  c:sdcm.results_analyze p:INFO  > Performance average of 2 results for: repair_runtime_large_partitions on version: 3.0.8 is: 111.119158387
16:51:41 < t:2019-08-28 13:51:41,839 f:tester.py       l:1505 c:sdcm.tester          p:ERROR > Failed to check regression: Current test performance for: repair_runtime_large_partitions exceeds allowed deviation (133.342990065) < t:2019-08-28 13:51:41,839 f:tester.py       l:1505 c:sdcm.tester          p:ERROR > Failed to check regression: Current test performance for: repair_runtime_large_partitions exceeds allowed deviation (133.342990065)
```
logs in: https://jenkins.scylladb.com/job/scylla-staging/job/row-level-repair/job/test_row_level_repair_large_partitions_3-0/label=aws-eu-west1-qa-builder1/7/console

File: test-cases/performance/perf-row-level-repair-1TB.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                    "cassandra-stress write no-warmup cl=QUORUM n=125000000 -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=200 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=375000001..500000000",
+                    "cassandra-stress write no-warmup cl=QUORUM n=125000000 -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=200 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=500000001..625000000",
+                    "cassandra-stress write no-warmup cl=QUORUM n=125000000 -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=200 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=625000001..750000000",
+                    "cassandra-stress write no-warmup cl=QUORUM n=125000000 -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=200 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=750000001..875000000",
+                    "cassandra-stress write no-warmup cl=QUORUM n=125000000 -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=200 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=875000001..1000000000"]
+
+n_db_nodes: 3
+n_loaders: 8
```
Comment (bentsi) [inline test-cases/performance/perf-row-level-repair-1TB.yaml]: 8 loader? is it expected?

File: test-cases/performance/perf-row-level-repair-1TB.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
+instance_type_loader: 'c4.2xlarge'
+instance_type_monitor: 't3.small'
```
Comment (bentsi) [inline test-cases/performance/perf-row-level-repair-1TB.yaml]: why did you change default (`instance_type_monitor: 't3.large'`)? less than this type it won't work for you

File: test-cases/performance/perf-row-level-repair-1TB.yaml (hunk 1 above)
Comment (yarongilor) [reply test-cases/performance/perf-row-level-repair-1TB.yaml]: yes, it saves time on prefill the 'prepare' data of test.

File: test-cases/performance/perf-row-level-repair-1TB.yaml (hunk 2 above)
Comment (yarongilor) [reply test-cases/performance/perf-row-level-repair-1TB.yaml]: t3.small is used on all perf-regression yamls.

File: performance_regression_row_level_repair_test.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+
+class PerformanceRegressionRowLevelRepairTest(ClusterTester):
+    """
+    Test Scylla performance regression with cassandra-stress.
+
+    """
+
+    KEYSPACE_NAME = "ks"
+    TABLE_NAME = "cf"
+    INT_COLUMNS = 20  # TODO: 99
```
Comment (fruch) [inline performance_regression_row_level_repair_test.py]: what are all those TODO ?, we want them or not ? if not, please remove them

File: performance_regression_row_level_repair_test.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+                self.get_stress_results(queue=stress, store_results=False)
+
+            self.update_test_details()
+        else:
+            self.log.warning("No prepare command defined in YAML!")
+
+    # Util functions ===============================================================================================
+
+
+    # Global Util functions ===============================================================================================
```
Comment (fruch) [inline performance_regression_row_level_repair_test.py]: what's the difference between `utils functions` to `global utils functions` ?

File: performance_regression_row_level_repair_test.py [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+        # if all are zeros the result will be False, otherwise there are still compactions
+        if results:
+            assert any([float(v[1]) for v in results[0]["values"]]) is False, \
+                "Waiting until all compactions settle down"
+
+    def _disable_hinted_handoff(self):
+
+        yaml_file = "/etc/scylla/scylla.yaml"
+        for node in self.db_cluster.nodes:  # disable hinted handoff
+            res = node.remoter.run('sudo echo hinted_handoff_enabled: false >> {}'.format(yaml_file))
```
Comment (fruch) [inline performance_regression_row_level_repair_test.py]: can we do this the same in with the configuration, i.e. with sed ?, this can end up having `hinted_handoff_enabled` written multiple times in scylla.yaml, and might cause confusion.

File: sdcm/utils/common.py [hunk 6]
```diff
@@ -1015,3 +1015,14 @@ def get_branched_ami(ami_version, region_name):
         return amis
     else:
         return amis[:1]
+
+
+def measure_time(func):
+    @wraps(func)
```
Comment (fruch) [inline sdcm/utils/common.py]: can you add a docstring to this function, to explain what it's doing ? (see `log_run_info`)

File: performance_regression_row_level_repair_test.py (hunk 4 above)
Comment (fruch) [reply performance_regression_row_level_repair_test.py]: Would love to see the those global one goes into tester.py, or to their own file

File: performance_regression_row_level_repair_test.py (hunk 3 above)
Comment (yarongilor) [reply performance_regression_row_level_repair_test.py]: removed.

File: test-cases/performance/perf-row-level-repair-1TB.yaml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+test_duration: 900
+
+prepare_write_cmd: ["cassandra-stress write no-warmup cl=QUORUM n=125000000 -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=200 -col 'size=FIXED(1024) n=FIXED(1)' -pop seq=1..125000000",
```
Comment (roydahan) [inline test-cases/performance/perf-row-level-repair-1TB.yaml]: Should be CL=ALL...

File: test-cases/performance/perf-row-level-repair-1TB.yaml (hunk 2 above)
Comment (roydahan) [reply test-cases/performance/perf-row-level-repair-1TB.yaml]: remove monitor instance.

File: test-cases/performance/perf-row-level-repair-1TB.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+user_prefix: 'perf-row-level-repair'
+failure_post_behavior: keep
+space_node_threshold: 644245094
+pre_create_schema: True
+
+round_robin: 'true'
+append_scylla_args: '--blocked-reactor-notify-ms 50'
+
+send_email: false
```
Comment (roydahan) [inline test-cases/performance/perf-row-level-repair-1TB.yaml]: change to true

File: test-cases/performance/perf-row-level-repair-1TB.yaml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+n_db_nodes: 3
+n_loaders: 8
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
+instance_type_loader: 'c4.2xlarge'
+instance_type_monitor: 't3.small'
+
+user_prefix: 'perf-row-level-repair'
+failure_post_behavior: keep
```
Comment (roydahan) [inline test-cases/performance/perf-row-level-repair-1TB.yaml]: change to destroy

File: test-cases/performance/perf-row-level-repair-1TB.yaml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+space_node_threshold: 644245094
```
Comment (roydahan) [inline test-cases/performance/perf-row-level-repair-1TB.yaml]: If you're not using this param, remove it from the yaml.

File: performance_regression_row_level_repair_test.py (hunk 4 above)
Comment (yarongilor) [reply performance_regression_row_level_repair_test.py]: moved some utils to tester.py . left only the 'private util functions'

File: performance_regression_row_level_repair_test.py (hunk 5 above)
Comment (yarongilor) [reply performance_regression_row_level_repair_test.py]: 'sed' is not good in case parameter doesn't exist in yaml. fixed this in another way with 'grep -v'.

File: sdcm/utils/common.py (hunk 6 above)
Comment (yarongilor) [reply sdcm/utils/common.py]: done

**Comment by yarongilor** [comment]
All comments are addressed. PR is ready for re-review.

**Comment by fruch** [comment]
@yarongilor please resolve conflicts, and make sure precommit is happy

@roydahan seems like all your comment were addressed, want to give it a quick last glance ?

**Comment by fruch** [comment]
@yarongilor  can you please add a pipeline for this one ?

**Comment by bentsi** [comment]
@yarongilor ?

**Comment by yarongilor** [comment]
> @yarongilor ?

unfortunately this PR was addresses 21 days after it was ready for review..

now it is too late and a rebase cannot be done. please change your review intervals so that PRs will not have to be re-written and re-written all over again for months.

**Comment by yarongilor** [comment]
> missing a pipeline

@fruch , Added jenkins-pipelines/perf-row-level-repair.jenkinsfile

**Comment by roydahan** [comment]
Let's merge this, it wasted too much of our time.
@yarongilor please make sure you have a jenkins job that is using this pipeline file and send me its link.

**Comment by bentsi** [comment]
pre-commit failed:
```
[2019-11-25T15:19:20.737Z] unit_tests/test_version_utils.py:1:0: R0801: Similar lines in 2 files

[2019-11-25T15:19:20.737Z] ==longevity_large_partition_test:16

[2019-11-25T15:19:20.737Z] ==performance_regression_row_level_repair_test:137

[2019-11-25T15:19:20.737Z]                     CREATE KEYSPACE IF NOT EXISTS scylla_bench WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 3}

[2019-11-25T15:19:20.737Z]             """)

[2019-11-25T15:19:20.737Z]             session.execute("""

[2019-11-25T15:19:20.737Z]                     CREATE TABLE IF NOT EXISTS scylla_bench.test (

[2019-11-25T15:19:20.737Z]                     pk bigint,

[2019-11-25T15:19:20.737Z]                     ck bigint,

[2019-11-25T15:19:20.737Z]                     v blob,

[2019-11-25T15:19:20.737Z]                     PRIMARY KEY (pk, ck)

[2019-11-25T15:19:20.737Z]                 ) WITH CLUSTERING ORDER BY (ck ASC)

[2019-11-25T15:19:20.737Z]                     AND bloom_filter_fp_chance = 0.01

[2019-11-25T15:19:20.737Z]                     AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}

[2019-11-25T15:19:20.737Z]                     AND comment = ''

[2019-11-25T15:19:20.737Z]                     AND compression = {}

[2019-11-25T15:19:20.737Z]                     AND crc_check_chance = 1.0

[2019-11-25T15:19:20.737Z]                     AND dclocal_read_repair_chance = 0.0

[2019-11-25T15:19:20.737Z]                     AND default_time_to_live = 0

[2019-11-25T15:19:20.737Z]                     AND gc_grace_seconds = 864000

[2019-11-25T15:19:20.737Z]                     AND max_index_interval = 2048

[2019-11-25T15:19:20.737Z]                     AND memtable_flush_period_in_ms = 0

[2019-11-25T15:19:20.737Z]                     AND min_index_interval = 128

[2019-11-25T15:19:20.737Z]                     AND read_repair_chance = 0.0

[2019-11-25T15:19:20.737Z]                     AND speculative_retry = 'NONE';

[2019-11-25T15:19:20.737Z]                             """) (duplicate-code)

[2019-11-25T15:19:20.737Z] 

[2019-11-25T15:19:20.737Z] -------------------------------------------------------------------

[2019-11-25T15:19:20.737Z] Your code has been rated at 9.99/10 (previous run: 10.00/10, -0.01)
```

**Comment by yarongilor** [comment]
Except for this minor pylint issue, test now runs OK in job:
https://jenkins.scylladb.com/job/scylla-staging/job/row-level-repair/job/perf-row-level-repair-large-partitions/13/console

@bentsi, please consider merge this PR per @roydahan 's request now. we can address the rest of pylint comments later on, thanks.

**Comment by bentsi** [comment]
your job failed:
```
 Failed to check regression: unsupported operand type(s) for +: 'int' and 'list' < t:2019-11-26  Failed to check regression: unsupported operand type(s) for +: 'int' and 'list'
 Traceback (most recent call last):
   File "/sct/sdcm/tester.py", line 1656, in check_specified_stats_regression
     perf_analyzer.check_regression(self._test_id, dict_specific_tested_stats=dict_specific_tested_stats, is_gce=is_gce)
   File "/sct/sdcm/results_analyze.py", line 303, in check_regression
     version_avg = sum(list_param_results) / float(len(list_param_results))
 TypeError: unsupported operand type(s) for +: 'int' and 'list'
```

**Comment by bentsi** [comment]
and why did you cherry-pick the commit "Get AWS keys from Jenkins in CI (#1474)" instead of rebasing on latest master?

**Comment by yarongilor** [comment]
> and why did you cherry-pick the commit "Get AWS keys from Jenkins in CI (#1474)" instead of rebasing on latest master?

just a temporarily fix. it is already addressed on a later rebase.

**Comment by yarongilor** [comment]
@bentsi , i opened issue: https://github.com/scylladb/scylla-cluster-tests/issues/1492
for the result analyze failure. the PR can be merged and the fix for this will be addressed in a later PR.

**Comment by yarongilor** [comment]
> @bentsi , i opened issue: #1492
> for the result analyze failure. the PR can be merged and the fix for this will be addressed in a later PR.
it may possibly be fixed by deleting a wrong value from Elastic-search db.

**Comment by yarongilor** [comment]
> > @bentsi , i opened issue: #1492
> > for the result analyze failure. the PR can be merged and the fix for this will be addressed in a later PR.
> 
> it may possibly be fixed by deleting a wrong value from Elastic-search db.

@bentsi , as you can see in log, the performance analyze did completed successfully, and reported correctly for this 3.1.0.rc2 version:
```
< t:2019-11-26 10:23:19,865 f:results_analyze.py l:299  c:sdcm.results_analyze p:INFO  > Performance result for: repair_runtime_large_partitions is: 50.586652755737305. (average statistics deviation limit is: 33.66557493209839
< t:2019-11-26 10:23:19,866 f:results_analyze.py l:305  c:sdcm.results_analyze p:INFO  > Performance average of 5 results for: repair_runtime_large_partitions on version: 3.1.0.rc2 is: 28.054645776748657
```

**Comment by fruch** [comment]
@yarongilor, still waiting for your updates on this PR.

**Comment by fruch** [comment]
@yarongilor, is this PR still relevant ?

**Comment by yarongilor** [comment]
> @yarongilor, is this PR still relevant ?

yes. perhaps will be addressed after 2019.1.4 release.

**Comment by roydahan** [comment]
Fix commits that aren't relevant to this PR, fix conflicts if lest.

**Comment by yarongilor** [comment]
> please rebase this PR and remove the unneeded AWS related commit.
> 
> Also please address the pylint commit, we can't merge without addressing those, it would effect all other PRs.

@fruch , PR is rebased and pylint should pass as well.

**Comment by fruch** [comment]
@yarongilor it should, but it didn't. You still have issues. from duplicated code to other issues.
Please run pre-commit locally and fix the issues.

**Comment by fruch** [comment]
@yarongilor I've open a different PR #1706 for the unit test code duplication issues

**Comment by fruch** [comment]
@yarongilor @roydahan, can you guys review the backport tags ? is that still need to backport to branch-3.1 ?

**Comment by roydahan** [comment]
No, changed to 3.2
